rate-limit/src/lib.rs

```diff
@@ -0,0 +1,454 @@
+//! # Rate Limit
+//! A module to provide rate limit for arbitrary type Key and integer type Value
+//!
+//! - [`Config`](./trait.Config.html)
+//! - [`Call`](./enum.Call.html)
+//! - [`Module`](./struct.Module.html)
+//!
+//! ## Overview
+//!
+//! This module is a utility to provide rate limiter for arbitrary type Key and
+//! integer type Value, which can config limit rule to produce quota and consume
+//! quota, and expose quota consuming checking and whitelist that can bypass
+//! checks.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::unused_unit)]
+
+use frame_support::{pallet_prelude::*, traits::UnixTime, transactional, BoundedVec};
+use frame_system::pallet_prelude::*;
+use orml_traits::{RateLimiter, RateLimiterError};
+use orml_utilities::OrderedSet;
+use parity_scale_codec::MaxEncodedLen;
+use scale_info::TypeInfo;
+use sp_runtime::traits::{BlockNumberProvider, SaturatedConversion, Zero};
+use sp_std::{prelude::*, vec::Vec};
+
+pub use module::*;
+pub use weights::WeightInfo;
+
+mod mock;
+mod tests;
+pub mod weights;
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	/// Period type.
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub enum Period {
+		Blocks(u64),
+		Seconds(u64),
+	}
+
+	/// Limit rules type.
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub enum RateLimitRule {
+		/// Each period to reset remainer quota to `quota` amount.
+		/// `can_consume` check return true when the remainer quota gte the
+		/// consume amount.
+		PerPeriod { period: Period, quota: u128 },
+		/// Each period to increase `quota_increment` amount to remainer quota
+		/// and keep remainer quota lte `max_quota`.
+		/// `can_consume` check return true when the remainer quota gte the
+		/// consume amount.
+		TokenBucket {
+			period: Period,
+			quota_increment: u128,
+			max_quota: u128,
+		},
+		/// can_consume check return true always.
+		Unlimited,
+		/// can_consume check return false always.
+		NotAllowed,
+	}
+
+	/// The maximum length of KeyFilter inner key.
+	pub const MAX_FILTER_KEY_LENGTH: u32 = 256;
+
+	/// Match rules to fitler key is in bypass whitelist.
+	#[derive(PartialOrd, Ord, PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+	pub enum KeyFilter {
+		/// If the encoded key is equal to the vec, the key is in whitelist.
+		Match(BoundedVec<u8, ConstU32<MAX_FILTER_KEY_LENGTH>>),
+		/// If the encoded key starts with the vec, the key is in whitelist.
+		StartsWith(BoundedVec<u8, ConstU32<MAX_FILTER_KEY_LENGTH>>),
+		/// If the encoded key ends with the vec, the key is in whitelist.
+		EndsWith(BoundedVec<u8, ConstU32<MAX_FILTER_KEY_LENGTH>>),
```

Comment:
I don't think we need this

Reply:
If the key is AccountId, `Match` pattern may usually be used, but `StartWith` pattern can still be used to match all the pallet accounts. Can consider retaining it, and should use `StartWith` and `EndWith` with caution in actual configuration.